### PR TITLE
adding backdoor attack code

### PR DESCRIPTION
**Pull request by Alan-Qin**

### CODE REVIEW

**Comment by joneswong** [comment]
@DavdGao @yxdyc According to current design of `Trainer`, what is your expected way to implement such a behavior that each client conducts evaluation on an additional dataset? Please have a look at @Alan-Qin  's workaround. Thanks.

File: federatedscope/attack/trainer/benign_trainer.py [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+    """
+
+    ctx['poison_'+ctx.cur_data_split+'_loader'] = ctx.data['poison_'+ctx.cur_data_split]
+    ctx['poison_'+ctx.cur_data_split+'_data'] = ctx.data['poison_'+ctx.cur_data_split].dataset
+    ctx['num_poison_'+ctx.cur_data_split+'_data'] = len(ctx.data['poison_'+ctx.cur_data_split].dataset)
+    setattr(ctx, "poison_{}_y_true".format(ctx.cur_data_split), [])
+    setattr(ctx, "poison_{}_y_prob".format(ctx.cur_data_split), [])
+    setattr(ctx, "poison_num_samples_{}".format(ctx.cur_data_split), 0)
+
+    for batch_idx, (samples, targets) in enumerate(ctx['poison_'+ctx.cur_data_split+'_loader']):
```
Comment (DavdGao) [inline federatedscope/attack/trainer/benign_trainer.py:181]: To conduct evaluation on the "poison" dataset for each client after a training routine, you can use the existing APIs as follows:
- Add the "poison" data into the global data dict, like
```
{
    '1': {
        "train": ...
        "test": ...
        "val": ...
        "posion": ...
    },
    '2': {
        "train": ...
        "test": ...
        "val": ...
        "posion": ...
    },
    ...
}
```
- Set `config.eval.split=["posion"]` to specifies the part of dataset for evaluation. If your choice metric is accuracy (`config.eval.metric=["acc"]`), then the evaluation results will be reported by the name of seleted split "posion" and metric "acc": "posion_acc"
- Set `config.eval.freq` as what you want.

By this way,  you don't need to code a test routine within the trainer and worry about duplicate variables within the context `ctx`.

File: federatedscope/attack/trainer/benign_trainer.py [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+
+    Args:
+        base_trainer: Type: core.trainers.GeneralTorchTrainer
+
+    :returns:
+        The wrapped trainer; Type: core.trainers.GeneralTorchTrainer
+
+    '''
+
+    # ---- action-level plug-in -------
```
Comment (DavdGao) [inline federatedscope/attack/trainer/benign_trainer.py:35]: It's suggested to remove the unused code.

**Comment by DavdGao** [comment]
> @DavdGao @yxdyc According to current design of `Trainer`, what is your expected way to implement such a behavior that each client conducts evaluation on an additional dataset? Please have a look at @Alan-Qin 's workaround. Thanks.

please see the inline comments

File: federatedscope/attack/auxiliary/backdoor_utils.py [hunk 3]
```diff
@@ -0,0 +1,450 @@
...
+    blend_img = np.clip(blend_img.astype('uint8'), 0, 255)
+
+    return blend_img
+
+
+def _signalTrigger( img, height, width, distance, trig_h, trig_w):
+    #  vertical stripe pattern different from sig
+    alpha = 0.2
+    # load signal mask
+    signal_mask = np.load('/mnt/zeyuqin/FederatedScope/federatedscope/attack/triggers/signal_cifar10_mask.npy')
```
Comment (Osier-Yi) [inline federatedscope/attack/auxiliary/backdoor_utils.py:246]: Please not hard code the file path. Please check all the file paths.

File: federatedscope/core/auxiliaries/trainer_builder.py [hunk 4]
```diff
@@ -131,12 +131,23 @@ def get_trainer(model=None,
...
+        trainer = wrap_benignTrainer(trainer)
+
     if is_attacker:
-        logger.info(
-            '---------------- This client is an attacker --------------------')
-        from federatedscope.attack.auxiliary.attack_trainer_builder import \
-            wrap_attacker_trainer
+        if 'backdoor' in config.attack.attack_method:
+            logger.info(
+                '---------------- This client is an backdoor attacker --------------------')
```
Comment (Osier-Yi) [inline federatedscope/core/auxiliaries/trainer_builder.py:141]: "an backdoor attacker" -> 'a backdoor attacker'

File: federatedscope/attack/auxiliary/backdoor_utils.py [hunk 5]
```diff
@@ -0,0 +1,450 @@
...
+    # img = (1-alpha) * np.uint32(img) + alpha * pattern
+    img = np.uint32(img) + pattern
+    img = np.uint8(np.clip(img, 0, 255))
+    return img
+
+
+def _wanetTrigger(img, height, width, distance, trig_w, trig_h, delta=20, f=6):
+    """
+    Implement paper:
+    """
```
Comment (Osier-Yi) [inline federatedscope/attack/auxiliary/backdoor_utils.py:331]: Missing paper name here.

File: federatedscope/attack/auxiliary/poisoning_data.py [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+                if i in perm:
+                    # select trigger
+                    # out_file = '/mnt/zeyuqin/FederatedScope/test_before_cifar.png'
+                    # # # matplotlib.image.imsave(out_file, np.squeeze(img, axis = -1), cmap = 'gray')
+                    # matplotlib.image.imsave(out_file, img)
+                    # import pdb; pdb.set_trace()
+                    img = selectTrigger(img, height, width, distance, trig_h, trig_w, trigger_type, load_path)
+                    # change target
+                    # out_file = '/mnt/zeyuqin/FederatedScope/test_after_cifar.png'
+                    # # matplotlib.image.imsave(out_file, np.squeeze(img, axis = -1), cmap = 'gray')
```
Comment (Osier-Yi) [inline federatedscope/attack/auxiliary/poisoning_data.py]: Please delete the unnecessary comments.

File: federatedscope/attack/auxiliary/create_edgeset.py [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+        images_seven_cut = images_seven_cut.unsqueeze(1)
+        print('size of images_seven_cut: ', images_seven_cut.size())
+        poisoned_labels_cut = (torch.zeros(images_seven_cut.size()[0]) +
+                               target_label).long()
+
+    else:
+        images_seven_DA = copy.deepcopy(images_seven)
+
+        cand_angles = [180 / fraction * i for i in range(1, fraction + 1)]
+        print("Candidate angles for DA: {}".format(cand_angles))
```
Comment (Osier-Yi) [inline federatedscope/attack/auxiliary/create_edgeset.py:56]: Please using logger.

File: federatedscope/contrib/model/resnet.py [hunk 8]
```diff
@@ -0,0 +1,305 @@
+from federatedscope.register import register_model
+'''Pre-activation ResNet in PyTorch.
+
+Reference:
+[1] Kaiming He, Xiangyu Zhang, Shaoqing Ren, Jian Sun
+    Identity Mappings in Deep Residual Networks. arXiv:1603.05027
+'''
```
Comment (Osier-Yi) [inline federatedscope/contrib/model/resnet.py:7]: Please add the corresponding licenses into FederatedScope/LICENSE

File: federatedscope/attack/auxiliary/poisoning_data.py [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
+logger = logging.getLogger(__name__)
+
+
+def load_poisoned_dataset_edgeset(data, ctx, mode):
+
+    transforms_funcs = get_transform(ctx, 'torchvision')['transform']
+    load_path = ctx.attack.edge_path
+    if "femnist" in ctx.data.type:
+        if mode == 'train':
+            fraction = 0.1
```
Comment (Osier-Yi) [inline federatedscope/attack/auxiliary/poisoning_data.py:25]: Is it a fixed value? Not user defined?

File: federatedscope/attack/auxiliary/poisoning_data.py [hunk 10]
```diff
@@ -0,0 +1,302 @@
...
+                poison_testset,
+                batch_size=ctx.data.batch_size,
+                shuffle=False,
+                num_workers=ctx.data.num_workers)
+
+    elif "CIFAR10" in ctx.data.type:
+        target_label = int(ctx.attack.target_label_ind)
+        target_label = 9
+        label = target_label
+        num_poisoned = 300
```
Comment (Osier-Yi) [inline federatedscope/attack/auxiliary/poisoning_data.py:60]: target label & num_poisoned are not user defined?

File: federatedscope/attack/auxiliary/poisoning_data.py [hunk 11]
```diff
@@ -0,0 +1,302 @@
...
+               surrogate_model=None,
+               load_path=None):
+
+    height = dataset[0][0].shape[-2]
+    width = dataset[0][0].shape[-1]
+    trig_h = int(trig_h * height)
+    trig_w = int(trig_w * width)
+
+    if 'wanet' in trigger_type:
+        cross_portion = 2
```
Comment (Osier-Yi) [inline federatedscope/attack/auxiliary/poisoning_data.py:123]: Is it a default val?

File: federatedscope/attack/auxiliary/poisoning_data.py [hunk 12]
```diff
@@ -0,0 +1,302 @@
...
+
+def load_poisoned_dataset_pixel(data, ctx, mode):
+
+    trigger_type = ctx.attack.trigger_type
+    label_type = ctx.attack.label_type
+    target_label = int(ctx.attack.target_label_ind)
+    transforms_funcs = get_transform(ctx, 'torchvision')['transform']
+
+    if "femnist" in ctx.data.type:
+        inject_portion_train = 0.2
```
Comment (Osier-Yi) [inline federatedscope/attack/auxiliary/poisoning_data.py:190]: Is it a fixed val?

File: federatedscope/contrib/model/resnet.py [hunk 13]
```diff
@@ -0,0 +1,305 @@
+from federatedscope.register import register_model
```
Comment (DavdGao) [inline federatedscope/contrib/model/resnet.py:1]: If the ResNet is the same with the one in torchvision.models, maybe we only need to modify `federatedscope/core/auxiliaries/model_builder.py` rather than copying it here.

File: federatedscope/attack/trainer/backdoor_trainer.py [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+
+
+def hook_on_fit_start_init_local_opt(ctx):
+
+    ctx.original_optimizer = ctx.optimizer
+    ctx.original_epoch = ctx["num_train_epoch"]
+    ctx["num_train_epoch"] = ctx.self_epoch
+    ctx.optimizer = torch.optim.SGD(
+        ctx.model.parameters(), lr=ctx.self_lr, momentum=0.0, weight_decay=0
+    )  # looks like adversary needs same lr to hide with others
```
Comment (Osier-Yi) [inline federatedscope/attack/trainer/backdoor_trainer.py:93]: Why the attacker's trainer need its own optimizer？

File: federatedscope/attack/auxiliary/create_edgeset.py [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+    fraction = 0.1
+    poisoned_edgeset = create_ardis_poisoned_dataset(fraction=fraction,
+                                                     data_path=data_path)
+
+    ardis_test_dataset = create_ardis_test_dataset(data_path)
+
+    print("Writing poison_data to: ")
+    print("poisoned_edgeset_fraction_{}".format(fraction))
+
+    with open("poisoned_edgeset_fraction_{}".format(fraction),
```
Comment (DavdGao) [inline federatedscope/attack/auxiliary/create_edgeset.py:136]: All the data processsing file are suggested to save in the directory `cfg.data.root`

File: federatedscope/attack/trainer/backdoor_trainer.py [hunk 16]
```diff
@@ -0,0 +1,181 @@
...
+
+
+def hook_on_fit_end_reset_opt(ctx):
+
+    ctx.optimizer = ctx.original_optimizer
```
Comment (Osier-Yi) [inline federatedscope/attack/trainer/backdoor_trainer.py:98]: In line 88: ctx.original_optimizer = ctx.optimizer and here ctx.optimizer = ctx.original_optimizer. It seems redundant?

File: federatedscope/attack/auxiliary/create_edgeset.py [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+              "wb") as saved_data_file:
+        torch.save(poisoned_edgeset, saved_data_file)
+
+    with open("ardis_test_dataset.pt", "wb") as ardis_data_file:
```
Comment (DavdGao) [inline federatedscope/attack/auxiliary/create_edgeset.py:140]: the same as above

File: federatedscope/attack/auxiliary/poisoning_data.py [hunk 18]
```diff
@@ -0,0 +1,302 @@
...
+import logging
+import os
+
+logger = logging.getLogger(__name__)
+
+
+def load_poisoned_dataset_edgeset(data, ctx, mode):
+
+    transforms_funcs = get_transform(ctx, 'torchvision')['transform']
+    load_path = ctx.attack.edge_path
```
Comment (DavdGao) [inline federatedscope/attack/auxiliary/poisoning_data.py:22]: Since the edge_path defaults to `''`, the poison data will be saved locally. 
Please save them in the directory `data.root`.

File: federatedscope/attack/auxiliary/poisoning_data.py [hunk 19]
```diff
@@ -0,0 +1,302 @@
...
+            for ii in range(num_poisoned_dataset):
+                sample = saved_southwest_dataset_train[ii]
+                data['train'].dataset.append((transforms_funcs(sample), label))
+
+            logger.info('adding {:d} edge-cased samples in CIFAR-10'.format(
+                num_poisoned))
+
+        if mode == 'test' or 'val':
+            poison_testset = list()
+            test_path = os.path.join(load_path,
```
Comment (DavdGao) [inline federatedscope/attack/auxiliary/poisoning_data.py:83]: The same as above, please save them in `data.root`

File: federatedscope/core/configs/cfg_attack.py [hunk 20]
```diff
@@ -13,6 +13,28 @@ def extend_attack_cfg(cfg):
...
     cfg.attack.attacker_id = -1
 
+    # for backdoor attack
+
+    cfg.attack.edge_path = ''
+    cfg.attack.trigger_path = ''
+    cfg.attack.setting = 'fix'
+    cfg.attack.freq = 10
+    cfg.attack.insert_round = 100000
+    cfg.attack.mean = [0.1307]
```
Comment (DavdGao) [inline federatedscope/core/configs/cfg_attack.py:23]: why does the mean and std default to 0.1307 and 0.3081? Maybe they should be set according to `cfg.data.type`？

**Review by DavdGao — COMMENTED** [review]
please see the inline comments.

**Comment by DavdGao** [comment]
@Alan-Qin  Please sign the contributor license agreement, thanks.

File: federatedscope/attack/auxiliary/poisoning_data.py [hunk 21]
```diff
@@ -0,0 +1,296 @@
...
+
+logger = logging.getLogger(__name__)
+
+
+def load_poisoned_dataset_edgeset(data, ctx, mode):
+
+    transforms_funcs = get_transform(ctx, 'torchvision')['transform']
+    load_path = ctx.attack.edge_path
+    if "femnist" in ctx.data.type:
+        if mode == 'train':
```
Comment (DavdGao) [inline federatedscope/attack/auxiliary/poisoning_data.py:24]: Please use the enums `MODE.TRAIN`, `MODE.TEST`... in `federatedscope/core/auxiliaries/enums.py` rather than use string directly.

Comment (DavdGao) [inline federatedscope/attack/auxiliary/poisoning_data.py:24]: Do we need to suport finetune (`mode` may equal to `MODE.FINETUNE`)?

**Comment by xieyxclack** [comment]
This thread is closed since the checking of Contributor License Agreement couldn't be passed, which is caused by the author of some commits couldn't be recognized as a normal GitHub user.

We have opened another pr (#267) to include the above modifications.

File: federatedscope/attack/auxiliary/create_edgeset.py [hunk 22]
```diff
@@ -0,0 +1,147 @@
...
+
+
+# creating the poisoned FEMNIST dataset with edge-case triggers
+
+def create_ardis_poisoned_dataset(base_label = 7, target_label = 1, fraction=0.1):
+
+    # we are going to label 7s from the ARDIS dataset as 1 (dirty label)
+
+    # load the data from csv's
+    ardis_images=np.loadtxt('/mnt/zeyuqin/FederatedScope/data/ARDIS/ARDIS_train_2828.csv', dtype='float')
```
[the remark this answers is not in the document]
Comment (Alan-Qin) [reply federatedscope/attack/auxiliary/create_edgeset.py]: fixed it

File: federatedscope/attack/trainer/benign_trainer.py (hunk 1 above)
Comment (Alan-Qin) [reply federatedscope/attack/trainer/benign_trainer.py]: fixed it

File: federatedscope/attack/trainer/benign_trainer.py (hunk 2 above)
Comment (Alan-Qin) [reply federatedscope/attack/trainer/benign_trainer.py]: fixed it

File: federatedscope/attack/auxiliary/backdoor_utils.py (hunk 3 above)
Comment (Alan-Qin) [reply federatedscope/attack/auxiliary/backdoor_utils.py]: fixed it

File: federatedscope/core/auxiliaries/trainer_builder.py (hunk 4 above)
Comment (Alan-Qin) [reply federatedscope/core/auxiliaries/trainer_builder.py]: fixed it

File: federatedscope/attack/auxiliary/backdoor_utils.py (hunk 5 above)
Comment (Alan-Qin) [reply federatedscope/attack/auxiliary/backdoor_utils.py:313]: fixed it

File: federatedscope/attack/auxiliary/poisoning_data.py (hunk 6 above)
Comment (Alan-Qin) [reply federatedscope/attack/auxiliary/poisoning_data.py]: fixed it

File: federatedscope/attack/auxiliary/create_edgeset.py (hunk 7 above)
Comment (Alan-Qin) [reply federatedscope/attack/auxiliary/create_edgeset.py]: fixed it

File: federatedscope/attack/auxiliary/poisoning_data.py (hunk 9 above)
Comment (Alan-Qin) [reply federatedscope/attack/auxiliary/poisoning_data.py]: fixed it

File: federatedscope/attack/auxiliary/poisoning_data.py (hunk 10 above)
Comment (Alan-Qin) [reply federatedscope/attack/auxiliary/poisoning_data.py]: fixed it

File: federatedscope/attack/auxiliary/poisoning_data.py (hunk 11 above)
Comment (Alan-Qin) [reply federatedscope/attack/auxiliary/poisoning_data.py]: fixed it

File: federatedscope/attack/auxiliary/poisoning_data.py (hunk 12 above)
Comment (Alan-Qin) [reply federatedscope/attack/auxiliary/poisoning_data.py]: fixed it

File: federatedscope/attack/trainer/backdoor_trainer.py (hunk 14 above)
Comment (Alan-Qin) [reply federatedscope/attack/trainer/backdoor_trainer.py]: fixed it

File: federatedscope/attack/trainer/backdoor_trainer.py (hunk 16 above)
Comment (Alan-Qin) [reply federatedscope/attack/trainer/backdoor_trainer.py]: fixed it

File: federatedscope/attack/auxiliary/create_edgeset.py [hunk 23]
```diff
@@ -0,0 +1,141 @@
...
+    images_seven = ardis_images[indices_seven, :]
+    images_seven = torch.tensor(images_seven).type(torch.uint8)
+
+    if fraction < 1:
+        num_sampled_data_points = (int)(fraction * images_seven.size()[0])
+        perm = torch.randperm(images_seven.size()[0])
+        idx = perm[:num_sampled_data_points]
+        images_seven_cut = images_seven[idx]
+        images_seven_cut = images_seven_cut.unsqueeze(1)
+        print('size of images_seven_cut: ', images_seven_cut.size())
```
[the remark this answers is not in the document]
Comment (Alan-Qin) [reply federatedscope/attack/auxiliary/create_edgeset.py]: fixed it

File: federatedscope/attack/auxiliary/create_edgeset.py [hunk 24]
```diff
@@ -0,0 +1,141 @@
...
+
+
+def create_ardis_poisoned_dataset(data_path=None,
+                                  base_label=7,
+                                  target_label=1,
+                                  fraction=0.1):
+    '''
+    creating the poisoned FEMNIST dataset with edge-case triggers
+    we are going to label 7s from the ARDIS dataset as 1 (dirty label)
+    load the data from csv's
```
[the remark this answers is not in the document]
Comment (Alan-Qin) [reply federatedscope/attack/auxiliary/create_edgeset.py:26]: fix it

File: federatedscope/attack/auxiliary/create_edgeset.py [hunk 25]
```diff
@@ -0,0 +1,141 @@
...
+    please use your own path of the data
+    '''
+    data_path = ''
+    fraction = 0.1
+    poisoned_edgeset = create_ardis_poisoned_dataset(fraction=fraction,
+                                                     data_path=data_path)
+
+    ardis_test_dataset = create_ardis_test_dataset(data_path)
+
+    print("Writing poison_data to: ")
```
[the remark this answers is not in the document]
Comment (Alan-Qin) [reply federatedscope/attack/auxiliary/create_edgeset.py]: fix it

File: federatedscope/attack/auxiliary/poisoning_data.py [hunk 26]
```diff
@@ -0,0 +1,302 @@
...
+    else:
+        data = load_poisoned_dataset_pixel(data, ctx, mode)
+    return data
+
+
+def poisoning(data, ctx):
+    for i in range(1, len(data) + 1):
+        if i == ctx.attack.attacker_id:
+            logger.info(50 * '-')
+            logger.info('start poisoning!!!!!!')
```
[the remark this answers is not in the document]
Comment (Alan-Qin) [reply federatedscope/attack/auxiliary/poisoning_data.py]: fix it

File: federatedscope/attack/auxiliary/create_edgeset.py (hunk 15 above)
Comment (Alan-Qin) [reply federatedscope/attack/auxiliary/create_edgeset.py:138]: fix it

File: federatedscope/attack/auxiliary/create_edgeset.py (hunk 17 above)
Comment (Alan-Qin) [reply federatedscope/attack/auxiliary/create_edgeset.py:142]: fix it

File: federatedscope/core/configs/cfg_attack.py (hunk 20 above)
Comment (Alan-Qin) [reply federatedscope/core/configs/cfg_attack.py:23]: yes

File: federatedscope/attack/trainer/backdoor_trainer.py [hunk 27]
```diff
@@ -0,0 +1,181 @@
...
+    ctx.original_model = copy.deepcopy(ctx.model)
+
+
+def hook_on_fit_end_scale_poisoning(ctx):
+
+    # conduct the scale poisoning
+    scale_para = ctx.scale_para
+
+    v = torch.nn.utils.parameters_to_vector(ctx.original_model.parameters())
+    logger.info("the Norm of the original global model: {}".format(
```
[the remark this answers is not in the document]
Comment (Alan-Qin) [reply federatedscope/attack/trainer/backdoor_trainer.py:109]: fix all

File: federatedscope/attack/auxiliary/poisoning_data.py (hunk 21 above)
Comment (Alan-Qin) [reply federatedscope/attack/auxiliary/poisoning_data.py:24]: fix
